apps/passport-client/src/service-worker.ts

```diff
@@ -1,65 +1,307 @@
-  );
-  await Promise.all(impermanentCache.map((item) => cache.delete(item)));
+/**
+ * Used to extract a key from a request, to compare it to the lists of
+ * cache resources above.  This may need updating if the list above ever
+ * includes more complex URLs, with parameters, or across domains.
+ */
+function requestToItemCacheKey(request: Request): string {
+  return new URL(request.url).pathname;
+}
 
-  await cache.addAll(resources);
+function logCacheName(cacheName: string): string {
+  return cacheName.substring(0, 9);
 }
 
-self.addEventListener("install", (event: any) => {
-  console.log(`[SERVICE_WORKER] installing ${process.env.SW_ID}`);
-  (self as any).skipWaiting();
+/**
+ * Helpers for quickly enabling or disabling verbose logging.
+ */
+const swLog = {
+  verbose: true, // TODO(artwyman): Disable before merging.
+  tag: `[SERVICE_WORKER][${process.env.SW_ID.substring(0, 4)}]`,
+  I: function (msg: string) {
+    console.log(this.tag, msg);
+  },
+  V: function (msg: string) {
+    if (this.verbose) {
+      console.debug(this.tag, msg);
+    }
+  }
+};
+
+/**
+ * Install is the first event which occurs when a new or updated ServiceWorker
+ * is registered.
+ */
+self.addEventListener("install", (event: ExtendableEvent) => {
+  swLog.V(`installing ${process.env.SW_ID}`);
+
+  if (!SERVICE_WORKER_ENABLED) {
+    return;
+  }
+
+  event.waitUntil(
+    (async () => {
+      // By default, the browser will only let one version of our ServiceWorker
+      // be active, and will wait to activate a new one until the old one isn't
+      // controlling any pages.  This call skips that so the new one can take
+      // over ASAP.
+      await swSelf.skipWaiting();
+
+      // Pre-populate our cache with all the resources we need to operate offline.
+      await prePopulateCaches();
+
+      swLog.I(`installed ${process.env.SW_ID}`);
+    })()
+  );
+});
+
+/**
+ * Activate happens after install, when the ServiceWorker is ready to take
+ * control of clients.
+ */
+self.addEventListener("activate", (event: ExtendableEvent) => {
+  // Clean up any stray service workers from prior development.
+  if (!SERVICE_WORKER_ENABLED) {
+    swLog.I(`self-unregistering`);
+    event.waitUntil(
+      (async () => {
+        await swSelf.registration.unregister();
+        for (const client of await swSelf.clients.matchAll()) {
+          (client as WindowClient).navigate(client.url);
+        }
+      })()
+    );
+    return;
+  }
+
+  swLog.V(`activating ${process.env.SW_ID}`);
 
   event.waitUntil(
-    addResourcesToCache([
-      ...impermanentCache,
-      "/semaphore-artifacts/16.wasm",
-      "/semaphore-artifacts/16.zkey",
-      "/fonts/IBMPlexSans-Regular.ttf",
-      "/fonts/IBMPlexSans-Medium.ttf",
-      "/fonts/IBMPlexSans-Light.ttf"
-    ])
+    (async () => {
+      // If supported, NavigationPreload allows fetches to start in
+      // parallel to the running of our fetch event handler.
```

Comment:
why do we want to do this?

Reply:
It makes the fetch faster when not served from the cache, because it's already started before the service worker runs, which could take a while if it boots up from idle.

See https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers#service_worker_navigation_preload